google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/readrows/ReadRowsResumptionStrategy.java

```diff
@@ -0,0 +1,182 @@
+  private ByteString lastKey = ByteString.EMPTY;
+  // Number of rows processed excluding Marker row.
+  private long numProcessed;
+
+  public ReadRowsResumptionStrategy(RowAdapter<RowT> rowAdapter) {
+    this.rowAdapter = rowAdapter;
+  }
+
+  @Override
+  public boolean canResume() {
+    return true;
+  }
+
+  @Override
+  public StreamResumptionStrategy<ReadRowsRequest, RowT> createNew() {
+    return new ReadRowsResumptionStrategy<>(rowAdapter);
+  }
+
+  @Override
+  public void onProgress(RowT response) {
+    // Last key can come from both the last processed row key and a synthetic row marker. The
+    // synthetic row marker is emitted when the server has read a lot of data that was filtered out.
+    // So it can trim the start of the scan, but does not contribute to the row limit.
+    lastKey = rowAdapter.getKey(response);
+    if (!rowAdapter.isScanMarkerRow(response)) {
+      // Only real rows count towards the rows limit.
+      numProcessed++;
+    }
+  }
+
+  /**
+   * {@inheritDoc}
+   *
+   * <p>Given a request, this implementation will narrow that request to exclude all row keys and
+   * ranges that would produce rows that come before {@link #lastKey}. Furthermore this
+   * implementation takes care to update the row limit of the request to account for all of the
+   * received rows.
+   */
+  @Override
+  public ReadRowsRequest getResumeRequest(ReadRowsRequest request) {
+    // An empty lastKey means that we have not successfully read the first row,
+    // resume with the original request object.
+    if (lastKey.isEmpty()) {
+      return request;
+    }
+
+    ReadRowsRequest originalRequest = request;
+
+    // Special case: empty query implies full table scan, make this explicit by adding an unbounded
+    // range to the request
+    if (request.getRows().getRowKeysList().isEmpty()
+        && request.getRows().getRowRangesList().isEmpty()) {
+
+      originalRequest =
+          request
+              .toBuilder()
+              .setRows(RowSet.newBuilder().addRowRanges(RowRange.getDefaultInstance()))
+              .build();
+    }
+
+    // Start building the resume request. The keys & ranges are cleared and will be recomputed.
+    Builder builder = originalRequest.toBuilder();
+    builder.clearRows();
+
+    RowSet.Builder rowSetBuilder = RowSet.newBuilder();
+
+    for (ByteString key : originalRequest.getRows().getRowKeysList()) {
+      if (ByteStringComparator.INSTANCE.compare(key, lastKey) > 0) {
+        rowSetBuilder.addRowKeys(key);
+      }
+    }
+
+    for (RowRange rowRange : originalRequest.getRows().getRowRangesList()) {
+      RowRange.Builder rowRangeBuilder = RowRange.newBuilder();
+
+      switch (rowRange.getEndKeyCase()) {
+        case END_KEY_CLOSED:
+          if (ByteStringComparator.INSTANCE.compare(rowRange.getEndKeyClosed(), lastKey) <= 0) {
+            continue;
+          } else {
+            rowRangeBuilder.setEndKeyClosed(rowRange.getEndKeyClosed());
+          }
```

Comment:
I would prefer consistent logical ordering for consistency. I prefer the order you used above (if compare(key, lastKey) > 0)) over the <=0 order here, because then the "active" case is the first block.